ai_edge_torch/generative/layers/experimental/ekv_cache.py

```diff
@@ -0,0 +1,178 @@
+      cls,
+      config: model_config.ModelConfig,
+      dtype: torch.dtype = torch.float32,
+      device: torch.device = None,
+  ) -> "KVCacheEntry":
+    shape = (
+        1,
+        config.kv_cache_max,
+        config.attn_config.num_query_groups,
+        config.head_dim,
+    )
+    k = torch.zeros(shape, dtype=dtype, device=device)
+    v = torch.zeros(shape, dtype=dtype, device=device)
+    obj = cls(k_cache=k, v_cache=v)
+    return obj
+
+
+@dataclass
+class EKVCache:
+  """A utility class for holding KV cach entries per layer."""
+
+  caches: Tuple[KVCacheEntry]
+
+  @classmethod
+  def from_model_config(
+      cls,
+      config: model_config.ModelConfig,
+      dtype: torch.dtype = torch.float32,
+      device: torch.device = None,
+  ) -> "EKVCache":
+    """Build an instance of the class based on model config.
+    Args:
+        config (ModelConfig): Model config used for building the cache.
+        dtype (torch.dtype, optional): The data type of the cache tensor.
+          Defaults to torch.float32.
+        device (torch.device, optional): The device placement of the cache
+          tensors. Defaults to None.
+    Returns:
+        EKVCache: The created cache object.
+    """
+    caches = [
+        KVCacheEntry.from_model_config(config, dtype, device)
+        for _ in range(config.num_layers)
+    ]
+    obj = cls(caches=tuple(caches))
+    return obj
+
+
+def _flatten_kvc(kvc: EKVCache) -> Tuple[List, List]:
+  flattened = []
+  flat_names = []
+  none_names = []
+  for i, kv_entry in enumerate(kvc.caches):
+    flattened.append(kv_entry.k_cache)
+    flat_names.append(f"k_{i}")
+    flattened.append(kv_entry.v_cache)
+    flat_names.append(f"v_{i}")
+  return flattened, [flat_names, none_names]
+
+
+def _flatten_kvc_with_keys(kvc: EKVCache) -> Tuple[List, List]:
+  flattened, (flat_names, none_names) = _flatten_kvc(kvc)
+  return [(pytree.MappingKey(k), v) for k, v in zip(flat_names, flattened)], flat_names
+
+
+def _unflatten_kvc(values: List[torch.Tensor], context: Tuple[List, List]) -> EKVCache:
+  assert len(values) % 2 == 0
+  num_layers = len(values) // 2
+  flat_names = context[0]
+  kv_entries = []
+  for i in range(num_layers):
+    k_cache_idx = flat_names.index(f"k_{i}")
+    v_cache_idx = flat_names.index(f"v_{i}")
+    kv_entries.append(
+        KVCacheEntry(k_cache=values[k_cache_idx], v_cache=values[v_cache_idx])
+    )
+  obj = EKVCache(tuple(kv_entries))
+  return obj
+
+
+pytree.register_pytree_node(
+    EKVCache,
+    _flatten_kvc,
+    _unflatten_kvc,
+    flatten_with_keys_fn=_flatten_kvc_with_keys,
+    serialized_type_name="",
+)
+
+
+def update(
+    cache: KVCacheEntry,
+    input_pos: torch.Tensor,
+    k_slice: torch.Tensor,
+    v_slice: torch.Tensor,
+    use_hlfb: bool = True,
+) -> KVCacheEntry:
+  """Out of place update of Cach buffer.
+
+  Args:
+      cache (KVCacheEntry): The original cache buffer.
+      input_pos (torch.Tensor): The update slice positions.
```

Comment:
is it int32 tensor?

Reply:
The update uses index_copy which uses long tensor. So still int64 type.